fix: fix transaction confirmation logic

marcnjaramillo opened the pull request:
#### Problem
Currently, one issue is that `sendAndConfirmTransaction` relies on `confirmTransaction` to determine whether a transaction has expired. This is an issue because `confirmTransaction` simply waits for 60 seconds rather than using the `lastValidBlockHeight` associated with a transaction's `blockhash`.

#### Summary of Changes
This PR attempts to move the transaction confirmation logic closer to a correct state by allowing the `Transaction` constructor to accept the entire results object from calling `getLatestBlockhash` rather than only the `blockhash`. This requires creating a new `TransactionCtorFields__DEPRECATED` to continue allowing the current implementation to be used while updating `TransactionCtorFields` to accept the updated values.

Because of the relationship between transactions and messages, this change also required a change to `Message` to likewise accept the `lastValidBlockHeight` that results from calling `getLatestBlockhash`.

Finally, we have updated `sendAndConfirmTransaction` to no longer use the current `confirmTransaction`. Instead, it compares the current `blockHeight` to `lastValidBlockHeight` while waiting for the transaction to be confirmed.

Fixes #23949
Fixes #24730
Closes #24757

steveluscher commented:
I’ll look at this soon, thanks for your patience!

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -172,8 +191,22 @@ export class Transaction {
+    if (!opts) {
+      return;
+    } else if (Object.prototype.hasOwnProperty.call(opts, 'recentBlockhash')) {
+      console.warn(
+        'Creating a Transaction with `recentBlockhash` is deprecated. ',
+      );
```
This is good, but this warning will be delivered at runtime. Let's instead deliver the warning at typecheck time! Check out the TypeScript playground link I dropped [here](https://github.com/solana-labs/solana/issues/23949#issuecomment-1090612996) for an example of how to make TypeScript literally cross out deprecated invocations of this constructor in your editor, with a helpful message about what to do instead.

marcnjaramillo replied:
Ah, I see. Thanks! I'll take another look and make sure to get this updated.

stellaw1 replied:
@steveluscher following your method makes the `Transaction` type unrecognizable outside of `transaction.ts`

<img width="1127" alt="image" src="https://user-images.githubusercontent.com/46392676/164345385-18464d7e-83c4-4f5e-8087-b8b33ed97372.png">

I'm a little uncertain of the best practice to approach this since TS does not allow multiple constructors.

steveluscher replied:
Interesting! Do you have this checked in somewhere that I can take a look at?

stellaw1 replied:
@steveluscher for sure, I pushed it [here](https://github.com/MLH-Fellowship/solana/tree/deprecate-transaction-ctor) !

steveluscher replied:
Solved it! My initial take was way too complicated. ([link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQgNgewMYGsAWBDAzmqBeKLYAJwEsA7AcwG4AoW0SKAFWI3KwyWFIXIGFgCYgDFSEOABMsAfQAiAUQAKAJQX8AgswVz8UAPIBbUsAA8rdp268BQ0eKlYANFADkcDMAhF4ydNjRXAD4oADIoAG9aKChiCCQIcmBfVEwcAH4ALlhEVICoAB8ocgBXODg6AF86BnBoCw4uHj5BYTEJaT0omI8vH1z-DOzumNGAIwG0tGyUwbRqKGjRnuxgADUMOFJJWYAJCFJKNGBs0sMxiGIl6qWEYDRLgGVgEoAzV6zCEgoaWhvaJAeLBYKD8PjkeLNciRJYyIHrTbbPYHI4nYolc6XQpQErkSQQV4UCCSOgxGRxBJJWZTGaTfJFUrlbG4-GEiEk+gxAD0XKgcgJRKg92gFC8xFeXGgr2EQoeUFeuOsfCFdSgAHcTLgMFAtkQoAhXvqAG6XRAYaRLJZcgBU1qgAAF8WAKZ5iUsoHaAAYUxLJOk4T1QNJQC6JKBOl1eSQAOnddqUcAg2GgYGwIM9vW8fr8NMiEz8uwCtILARc8I2Wx2A32h2OpwxF2IlUDFCISZjcZYaFIII1TIu4YgzvirskUAoUG1AA1owAGOdQE3ELA2fWG+1YBAecgYLlqiBjADM0YAVlhY6NrVylkg+ERiCVuMIABQIMDALCfFhsRpK2xtBxpHkZQ1E0bQ5AASlJRZuV5Zg5QhNVCEOHcXjiG87xIR87Ffd9P2yBorChVp7A6LAoM5KAeS7MNWUFYVx0MMBE0MX1PFXFdKFQkp0JiW8OCwp9iFwj9PkIpobBI9pHGxcS-ykwDZEUVR1C0HQIJhZYYlIQ1nwAQjfD8NJGLTRjiNDyGg5ZKndUYdKgETzzSfQ1XIJRiDfS5QGfVwfSpf1Agg4zbOWaiAElgFcEEGK3SQAFoiBARN9XfGxzxC0Z+M3RNozVDBiHIHz+DidiqEnKBPTBcgIT-QMNXuCq-OzPIA3HEEIxHKNo261wKNMvjMP1KR9DwvRDJBbBv0sCSWjsaSgOU0C1LkKytPuHto3JeJfWpfICFikaP2jJrdpwVaYkqKAJCwaATK08LIuihCICQxLksMtKL36rLgGKF7DpBfbRsmuTiLmxTztGdbzzhVYKyRasUWOPREIB6NM36HMAnSdG4cRKsCyR4BIZiaHNpOgKUf+vDcb6Zq5hx-MWvmDKbNM64-noH6hskKqaqhKmkL5yEbGfCIKaxnBslcDAxiQVxKgo7nEOFv9BdBcERb4MWMfp3MIiZuZpdl+WyzxytkVrNEAEZZztypFeoIA))

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -172,8 +191,22 @@ export class Transaction {
+        'Creating a Transaction with `recentBlockhash` is deprecated. ',
+      );
+      const oldOpts = opts as TransactionCtorFields_DEPRECATED;
+      Object.assign(this, oldOpts);
+      this.recentBlockhash = oldOpts.recentBlockhash;
```
The first line implies the second line, doesn't it?

marcnjaramillo replied:
You know, I can't remember now why we had this set up like this. I think this was a leftover from something else we were trying and didn't remove. I'll update this!

steveluscher commented on web3.js/src/message.ts:
```diff
@@ -73,7 +80,15 @@ export class Message {
     this.accountKeys = args.accountKeys.map(account => new PublicKey(account));
-    this.recentBlockhash = args.recentBlockhash;
+    if (
+      Object.prototype.hasOwnProperty.call(args, 'latestBlockhash') &&
```
Wherever this `hasOwnProperty` check appears, I think you can just write it as `args.hasOwnProperty(...)`, right?

marcnjaramillo replied:
This one is an interesting one. If we use `args.hasOwnProperty(...)` TypeScript complains: 
```Do not access Object.prototype method 'hasOwnProperty' from target object.```

We could have disabled the rule that throws the error, but we weren't sure if that was really the best option. After doing some research on the error, we found an article that explained what the error was and how to address it -- which is what we used.

steveluscher replied:
Oh interesting. I've never seen this ESLint rule. https://eslint.org/docs/rules/no-prototype-builtins

I'm going to consider turning that off in our codebase.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -680,9 +713,13 @@ export class Transaction {
   static populate(
     message: Message,
     signatures: Array<string> = [],
+    lastValidBlockHeight?: LastValidBlockHeight,
   ): Transaction {
     const transaction = new Transaction();
     transaction.recentBlockhash = message.recentBlockhash;
+    if (lastValidBlockHeight) {
+      transaction.lastValidBlockHeight = lastValidBlockHeight;
+    }
```
Why do you need to do this again? Isn't this value now part of the `Message` type?

marcnjaramillo replied:
Hmm, yeah. I'll update this. Thanks!

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
+        return;
+      }
+      await sleep(500);
+      await checkBlockHeight();
```
This recursive approach – barring the presence of tail call optimization – will keep adding to the JavaScript call stack every time you recurse. Since the recursion here is only used for _repetition_, it's better to keep the stack shallow by using a loop.

marcnjaramillo replied:
Refactored this a bit to use `onSignature(...)` along with a `while` loop.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
+    if (blockHeight > transaction.lastValidBlockHeight!) {
+      throw new Error('Transaction has expired.');
+    } else {
+      if (status) {
```
Unless I've missed it, I don't think we ever ask for the signature status again here, right? This just recurses until the block height is exceeded, whether or not the signature ever gets confirmed?

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
-      options && options.commitment,
-    )
-  ).value;
+  const status = await connection.getSignatureStatus(signature);
```
Rather than to check the signature status in a loop, did you consider making a _subscription_ using `onSignature(…)`? That way, the client can chill out and wait for the server to _push_ updates to the status, if any, while it keeps polling for the latest block height.

marcnjaramillo replied:
We noticed that this was being done in the original `confirmTransaction` and we did try to implement the same logic here, but it got really messy with all of the imports we needed so we abandoned it for the time being. We can definitely explore it again.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
-    throw new Error(
-      `Transaction ${signature} failed (${JSON.stringify(status)})`,
+  const checkBlockHeight = async () => {
+    const blockHeight = await connection.getBlockHeight(
+      options && options.commitment,
     );
```
Whatever we do, we need a way to be able to ignore the result of this function in case we learn, in the meantime, that the signature has confirmed. You never want to have learned that the signature confirmed, but _also_ be waiting on this RPC call to complete before notifying your application.

Maybe you could accomplish this with _two_ long-running promises, each with their own `AbortController`, and `Promise.race`.

marcnjaramillo replied:
We talked about this for a bit last night, and I think we understand what you're asking for but we haven't gotten anything concrete complete.

marcnjaramillo replied:
I spent some time working on this tonight, and I think I have something closer to what you had in mind. Wouldn't mind another quick look to see what you think of the updates.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
+      if (status) {
+        return;
+      }
+      await sleep(500);
```
I wonder if we can use the knowledge of how far away we are from `lastValidBlockHeight` to minimize network chatter. Could you wait to poll this until we get close, then poll faster and faster as the last valid height approaches?

steveluscher reviewed (changes requested):
This is going to be so great when it's fixed. Thanks for working on it! We have a good deal of ground to cover here before this is fast, light, and reliable enough for production, but start with these comments and we'll get there soon!

steveluscher commented on web3.js/src/util/promise-expiration.ts:
```diff
@@ -0,0 +1,14 @@
+export function promiseExpiration<T>(
+  promise: Promise<T>,
+  blockHeight: () => void,
+): Promise<T | null> {
+  let blockId: () => void;
+  const isExpired: Promise<null> = new Promise(resolve => {
+    (blockId = () => resolve(null)), blockHeight;
+  });
+
+  return Promise.race([promise, isExpired]).then((result: T | null) => {
+    blockId;
+    return result;
+  });
+}
```
I think you don't need this if the suggestion I made elsewhere in this review works!

steveluscher reviewed (commented):
Great changes; keep going!

marcnjaramillo commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -34,36 +40,64 @@ export async function sendAndConfirmTransaction(
-      await checkBlockHeight();
+      currentBlockHeight = await checkBlockHeight();
     }
   };
```
That's what we originally set out to do, but it may not have been the best approach. We made the recursive call and then exited once `getSignatureStatus` was successful, otherwise it would throw an error if the `blockHeight` exceeded `lastValidBlockHeight`. For this refactor, we decided to make it so this function was only responsible for monitoring where the `blockHeight` is in relation to `lastValidBlockHeight`, passing both of these to the `promiseExpiration` function. Looking at this now, I see I forgot to do something in the event the loop is exited, which should have been something about letting the user know the transaction expired. Taking all of your suggestions into account, I think this can be cleaned up quite a bit.

steveluscher replied:
Yeah! It's totally fine for the _last_ fetch to finish, just don't start new ones once we get a signature confirmation.

marcnjaramillo commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -34,36 +40,64 @@ export async function sendAndConfirmTransaction(
+    return blockHeight;
+  };
+
+  let currentBlockHeight = await checkBlockHeight();
```
That makes sense. I should have done that originally. Thanks for pointing it out!

marcnjaramillo commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -34,36 +40,64 @@ export async function sendAndConfirmTransaction(
-  //   )
-  // ).value;
+  try {
+    await promiseExpiration(confirmTx, expireTx);
```
I tried to implement this, but it was tricky because `expireTx` uses `async/await`, so converting it to `Promise` syntax proved to be challenging. I think what I have is correct. I'd like your feedback on `expireTx` specifically.

steveluscher replied:
Right. There are a couple of ways around this. All of these are equivalent. You can see how you can translate between them.

```
const result = await new Promise(resolve => {
  getResult().then(resolve);
});
```

```
const result = await new Promise(async resolve => {
  const result = await getResult();
  resolve(result);
});
```

```
const result = await new Promise(resolve => {
  (async () => {
    const result = await getResult();
    resolve(result);
  })();
});
```

```
const result = await (async {
  return await getResult();
})();
```

In all four of these cases, you can read and write to a `let done` binding that lives _outside_ of them. This is how you can cancel your `while` loop if you know the transaction to have completed.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
+  }
+
+  if (status === null) {
+    throw new Error('Transaction has expired.');
```
* Include the signature in this error message.
* I'd prefer if we started to throw first-class exceptions. That way people can catch them and do `instanceof` checks on them.

```
class TransactionExpiredBlockheightExceededError extends Error {
  constructor(signature) {
    super(`Signature: ${signature}`);
    this.signature = signature;
  }
}
Object.defineProperty(TransactionExpiredBlockheightExceededError.prototype, 'name', {
  value: 'TransactionExpiredBlockheightExceededError',
});

if (...) {
  throw new TransactionExpiredBlockheightExceededError(signature);
}
```

marcnjaramillo replied:
Okay, I'll work on this. Thanks!

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -172,8 +191,22 @@ export class Transaction {
    */
-  constructor(opts?: TransactionCtorFields) {
-    opts && Object.assign(this, opts);
+  constructor(opts?: TransactionCtorFields | TransactionCtorFields_DEPRECATED) {
```
Still needs deprecation using TypeScript method overloads.

marcnjaramillo replied:
Stella was working on this, so I'll check in with her to see how it's going.

stellaw1 replied:
https://github.com/solana-labs/solana/pull/24211#discussion_r854668811

steveluscher replied:
Solved it! My initial take was way too complicated. ([link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQgNgewMYGsAWBDAzmqBeKLYAJwEsA7AcwG4AoW0SKAFWI3KwyWFIXIGFgCYgDFSEOABMsAfQAiAUQAKAJQX8AgswVz8UAPIBbUsAA8rdp268BQ0eKlYANFADkcDMAhF4ydNjRXAD4oADIoAG9aKChiCCQIcmBfVEwcAH4ALlhEVICoAB8ocgBXODg6AF86BnBoCw4uHj5BYTEJaT0omI8vH1z-DOzumNGAIwG0tGyUwbRqKGjRnuxgADUMOFJJWYAJCFJKNGBs0sMxiGIl6qWEYDRLgGVgEoAzV6zCEgoaWhvaJAeLBYKD8PjkeLNciRJYyIHrTbbPYHI4nYolc6XQpQErkSQQV4UCCSOgxGRxBJJWZTGaTfJFUrlbG4-GEiEk+gxAD0XKgcgJRKg92gFC8xFeXGgr2EQoeUFeuOsfCFdSgAHcTLgMFAtkQoAhXvqAG6XRAYaRLJZcgBU1qgAAF8WAKZ5iUsoHaAAYUxLJOk4T1QNJQC6JKBOl1eSQAOnddqUcAg2GgYGwIM9vW8fr8NMiEz8uwCtILARc8I2Wx2A32h2OpwxF2IlUDFCISZjcZYaFIII1TIu4YgzvirskUAoUG1AA1owAGOdQE3ELA2fWG+1YBAecgYLlqiBjADM0YAVlhY6NrVylkg+ERiCVuMIABQIMDALCfFhsRpK2xtBxpHkZQ1E0bQ5AASlJRZuV5Zg5QhNVCEOHcXjiG87xIR87Ffd9P2yBorChVp7A6LAoM5KAeS7MNWUFYVx0MMBE0MX1PFXFdKFQkp0JiW8OCwp9iFwj9PkIpobBI9pHGxcS-ykwDZEUVR1C0HQIJhZYYlIQ1nwAQjfD8NJGLTRjiNDyGg5ZKndUYdKgETzzSfQ1XIJRiDfS5QGfVwfSpf1Agg4zbOWaiAElgFcEEGK3SQAFoiBARN9XfGxzxC0Z+M3RNozVDBiHIHz+DidiqEnKBPTBcgIT-QMNXuCq-OzPIA3HEEIxHKNo261wKNMvjMP1KR9DwvRDJBbBv0sCSWjsaSgOU0C1LkKytPuHto3JeJfWpfICFikaP2jJrdpwVaYkqKAJCwaATK08LIuihCICQxLksMtKL36rLgGKF7DpBfbRsmuTiLmxTztGdbzzhVYKyRasUWOPREIB6NM36HMAnSdG4cRKsCyR4BIZiaHNpOgKUf+vDcb6Zq5hx-MWvmDKbNM64-noH6hskKqaqhKmkL5yEbGfCIKaxnBslcDAxiQVxKgo7nEOFv9BdBcERb4MWMfp3MIiZuZpdl+WyzxytkVrNEAEZZztypFeoIA))

marcnjaramillo replied:
Stella pushed some changes for this. I also added a separate constructor overload for `nonce` transactions to address your [comment](https://github.com/solana-labs/solana/issues/23949#issuecomment-1092221677) about transactions being constructed with either `nonce` or `latestBlockhash`. It seemed to be the cleanest way to do it. What are your thoughts?

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
+        signature,
+        (result: SignatureResult, context: Context) => {
+          subscriptionId = undefined;
+          status = {
```
You never actually use this except to check if it's non-null. Is there a reason why we can't just eliminate `status` and have a `succeeded` flag instead?

marcnjaramillo replied:
Okay, after looking into this more I think I understand how to use this appropriately. I kept it as is but moved it back into `confirmTransaction` and changed `status` back to `response`. Then I check for `err` (which is on `SignatureResult`) like in the original implementation.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
+            context,
+            value: result,
+          };
+          message = 'confirmed';
```
👍

steveluscher replied:
Use TypeScript enums here.

```
enum TransactionStatus {
  CONFIRMED,
  EXPIRED,
}
```

steveluscher reviewed (dismissed):
Keep on going!

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
+    options?.preflightCommitment || options?.commitment;
+  let subscriptionId: number | undefined;
+  let status: RpcResponseAndContext<SignatureResult> | null = null;
+  let message: string | null = null;
```
Since you're returning your status code from your promises now, you don't need this.

marcnjaramillo commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -33,17 +39,81 @@ export async function sendAndConfirmTransaction(
   );
 
-  const status = (
-    await connection.confirmTransaction(
-      signature,
-      options && options.commitment,
-    )
```
Yes, that's the goal. I saw your comment about teaching `confirmTransaction` to use `latesetBlockhash` so that was the direction we wanted to take. I'll get this updated to reflect that change.

marcnjaramillo replied:
Okay, I went ahead and started working on this. I'm still not used to working with overload functions in TypeScript, so I did run into some issues. From what I could find while researching this, TypeScript only allows function overloading if the number of arguments are the same, which means I couldn't use two separate signatures that each had a different number of arguments. Instead, I added a new method `confirmTransactionBlockHeight` that uses the new approach and deprecated `confirmTransaction`. Then, in `sendAndConfirmTransaction` I added a variable to hold `recentBlockhash` and `lastValidBlockHeight` to use with the new method. I was looking at `sendTransaction` and I noticed that `_blockhashInfo` only uses `recentBlockhash`. Do you think we should update these as well?

marcnjaramillo replied:
I managed to figure this out. I had to disable the `no-dupe-class-members` eslint rule, and I wasn't getting issues. I tend not to mess with eslint rules when it's not my codebase, so if there's a reason this rule shouldn't be disabled I left the other implementation in (commented out). They're both there for comparison, so let me know what you think is best.

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2761,80 +2762,257 @@ export class Connection {
-    } catch (err) {
-      throw new Error('signature must be base58 encoded: ' + signature);
-    }
+    if (typeof signature === 'string') {
```
This would also let you delete `promise-timeout.ts` from the codebase completely.

marcnjaramillo replied:
Oh, interesting. I hadn't thought of doing it this way. I will certainly see if I can extract the logic from `promise-timeout` and adjust it so it lines up with the other strategy. I'll keep you posted!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2761,80 +2762,257 @@ export class Connection {
+      const expireTx = new Promise((resolve, reject) => {
+        (async () => {
+          try {
+            let currentBlockHeight = await checkBlockHeight();
```
You might be done by the time this finishes.
```suggestion
            let currentBlockHeight = await checkBlockHeight();
            if (done) return;
```

marcnjaramillo replied:
Thanks, I'll adjust this. I knew this could be improved somehow!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2761,80 +2762,257 @@ export class Connection {
+              currentBlockHeight <=
+              signature.latestBlockhash.lastValidBlockHeight!
+            ) {
+              if (done) return;
+              await sleep(1000);
+              currentBlockHeight = await checkBlockHeight();
```
You only need to check for doneness after each asynchronous step.
```suggestion
              await sleep(1000);
              if (done) return;
              currentBlockHeight = await checkBlockHeight();
              if (done) return;
```

marcnjaramillo replied:
I'll make this change. Thanks!

steveluscher reviewed (dismissed):
So close! I can see it coming together.

It's time to start writing tests for this stuff. Are you familiar with writing tests in this codebase, or should we hop on a call to talk through it?

steveluscher commented:
You haven't rebased atop the main branch in a _really_ long time. You might want to `git pull --rebase git@github.com:solana-labs/solana.git master` sooner than later. I just tried it and your PR applied cleanly.

steveluscher commented:
@stellaw1, when you rebase this code atop the main branch again, you'll discover a build error caused by the fact that these two definitions snuck in between your constructor overloads and the constructor implementation.

<img width="393" alt="image" src="https://user-images.githubusercontent.com/13243/165409215-a2ded7bb-f3db-4e90-add5-052814f5fe5d.png">

That's what's causing the build errors on this PR. Just move them out of the way so that all 4 constructor signatures come one after another.

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-dupe-class-members */
```
> You haven't rebased atop the main branch in a _really_ long time. You might want to `git pull --rebase git@github.com:solana-labs/solana.git master` sooner than later. I just tried it and your PR applied cleanly.

Yeah, I'm a bit embarrassed that I hadn't. When Stella pushed her changes I went ahead and did a rebase on my master branch and then made sure my feature branch was also up to date. It was a bit messy at first, but I think I managed to get everything straightened out. I try to rebase more often, but I lost track of time.

marcnjaramillo replied:
I haven't tried it, but I think it should work. I'll make the change and check it out!

marcnjaramillo commented:
> You haven't rebased atop the main branch in a _really_ long time. You might want to `git pull --rebase git@github.com:solana-labs/solana.git master` sooner than later. I just tried it and your PR applied cleanly.

Yeah, I'm a bit embarrassed that I hadn't rebased in so long. When Stella pushed her changes I did a rebase to get everything up to date. I try to rebase more frequently, but I just lost track of time and forgot about it. I'll make sure to keep on it now!

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2761,80 +2762,257 @@ export class Connection {
+  ): Promise<RpcResponseAndContext<SignatureResult>>;
+
+  async confirmTransaction(
+    signature: any,
```
I think this is reasonable. I had `any` because the examples for writing overloads I had found all said that the types on the implementation needed to be as generic as possible to be able to match the overloads. But in the `Transaction` constructor overload we had it as a union of possible types and that worked. I'll get this changed.

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2761,80 +2762,257 @@ export class Connection {
+  // }
+
+  confirmTransaction(
+    signature: {
+      signature: TransactionSignature;
+      latestBlockhash: TransactionBlockhash;
+    },
```
I agree. I'm not too sure about how to do this. I think I understand the basic idea of `nonce`s but I haven't actually used one before. This was the next thing I wanted to figure out.

steveluscher replied:
I'd say just make sure that you're not painting yourself into a corner. For instance, there's a potential footgun here with the type system once you add an `opts` type for nonce-based confirmations: ([playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9gOwMYQMKIGYEsDmUC8UA3gFBRQKIoBcUAzsAE5YI4A0JAvgNwkmiQoAZVwIAhsACujNJlwFiZeqInSItBs1YdyAGwkQGAIV1wkAawAWYupdoJJAWwBGERhx58Mk5MCyIoJDlGRwAVRjEEOjEkP0QACgiAd3QEbBxaeGRZNPkAH2EVKRlU9IBKRXIsDCh4gHlnACsIWIA6MEY4YC7wCFbrOjqkhAAFTshGUFakMV1dRLEUuXYoAHJKbNWyitJyciCo4EDlhWTS+RtYKhz03j2oAHoHqDGIMDEZCmuoYEsWHDorSUnCgEF0dGguz2BwYx1yeEIZxOlxEOHExRuuDueyeLxk70+ADdZlgACZQZymCyWCC4SxHX7-QHArh8A7YELhSLRWL+BDxIgbGhrDBwOCrThlbiPZ4AcTFpJI7KwnIiURicX5RDoRTUtFWovFbCg+mAhmAJjMVhsdigAEYAEwAZkl0tx8rgiqVwTCap5moFOrRqhk+sNq2NQvUIrFEqlMuEljgkl0pIQqyOrigs1MSQgpIAhEA))

You'll need a type strategy to make sure that people can't supply incorrect config.

One way of doing so is to make the config a tagged union, but then you'll likely need helpers to create these tagged objects, so that people don't have to muck around and do it themselves. Adds a layer of complexity that we maybe don't want. ([playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9gOwMYQMKIGYEsDmUC8UA3gFBRQD6FokAXFAOQKIoC0ARgIYDOEAJqySZcDADRkozZBHrdgAJywIc4gL4BuEiRrQAyrgSdgAV3lpheQqXJUd9BuwA2cJAGsAFhFzvgHHv0ELMQluAyNTGSg5RWVxckcjCDkAIWc3dx53egRjAFt2CHk1TRIMY2RgLEQoIQRseVyAFXlOBG5OJErEAAoWgHd0Otx6eGlB7DwAHyh9HEMTM3HcAEpiCSwMKF7OAYsAOltwaHwTxik2Ll4BWomGVetyGsQ5J6HLKH6lnE1HqAB6P5QAAKZjAnDMkhY0GA7iUOG4ewkqigEEcvDWv1qLxuuAIHx2Xx+jwBwNB4OgADdOI4sHwoE4XB4vDgfFAYXCEUiSKotDiGs1Wu1OlUEN0kGZEqMUF9ugwMHA4Hdlup-oCAOIKvgkPlNFptDpdUXiiCJVKMzzeYDJfx8GVyhViKAARgATABmZbK1VQDVwLXa4T8vVCw3dIihObhMz2eWK0SQ6Qxh2qL0k3TuODGRx8BAMYD06DU5x9fgAQi0ZQqIpqEuAECl5je3XOkWicOWIyhXwxHwgCwQPZs1CO9hbfiugTewV+LbiUA03Ir5WF1WNprSTMt1quMoj8wisgUcPjCTrKQ3GW4WUkeQK8g7s33iwsPbM-cHlGHdEYDPSzJ844BDiODTo8p5JFaF6ZHO5B7lGEBzguqhAA))

marcnjaramillo commented:
> @stellaw1, when you rebase this code atop the main branch again, you'll discover a build error caused by the fact that these two definitions snuck in between your constructor overloads and the constructor implementation.
> 
> <img alt="image" width="393" src="https://user-images.githubusercontent.com/13243/165409215-a2ded7bb-f3db-4e90-add5-052814f5fe5d.png">
> 
> That's what's causing the build errors on this PR. Just move them out of the way so that all 4 constructor signatures come one after another.

Yeah...this is probably my fault. I added the extra constructor overload and changed the deprecated one. [(comment)](https://github.com/solana-labs/solana/pull/24211#discussion_r859049408)

marcnjaramillo commented:
> So close! I can see it coming together.
> 
> It's time to start writing tests for this stuff. Are you familiar with writing tests in this codebase, or should we hop on a call to talk through it?

I haven't done much with tests in this codebase, so if you've got some time to get on a call I would appreciate it.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -17,6 +17,22 @@ import type {CompiledInstruction} from './message';
+  lastValidBlockHeight?: number;
+};
+
+export type TransactionSignatureBlockhash = {
+  signature: TransactionSignature;
+  blockhash?: Blockhash;
+  lastValidBlockHeight?: number;
+};
```
1. You can intersect these two types to make this one (see the `&` operator below)!
2. I'm not sure this belongs in the `Transaction` file. This is more like the shape of the _configuration_ specific to the `confirmTransaction` method. It should live alongside that.

Think of what you're doing as _refactoring_ the `confirmTransaction` method to take a config instead of a signature string. That config can be one of:

* A nonce-based config
* A blockheight-based config
* A signature string (**deprecated**)

If you were to write this from scratch it would look something like this:

```
/** @deprecated Instead, call `confirmTransaction` using a `TransactionConfirmationConfig` */
function confirmTransaction(config: TransactionSignature);
function confirmTransaction(config: BlockheightBasedTransactionConfimationConfig);
function confirmTransaction(config: NonceBasedTransactionConfirmationConfig);
function confirmTransaction(
  config:
    | BlockheightBasedTransactionConfimationConfig
    | NonceBasedTransactionConfirmationConfig
) {
  /* ... */
}
```

Then, you need types that describe the config:

```
type BlockheightBasedTransactionConfimationConfig = {
  signature: TransactionSignature,
} & TransactionBlockhash;
type NonceBasedTransactionConfirmationConfig = {...};
```

Another, maybe better name for these types would be ‘strategy,’ since that's what it _really_ is.

```
function confirmTransaction(strategy: BlockheightBasedConfirmationStrategy | NonceBasedConfirmationStrategy) { ... }
```

Yeah. That makes more sense, to me. What do you think about that?

marcnjaramillo replied:
> I'm not sure this belongs in the Transaction file. This is more like the shape of the configuration specific to the confirmTransaction method. It should live alongside that.

I wanted to try to keep things clean, and since we already had `TransactionSignature` defined in `Transaction` it seemed the best place to put it, but I also see your point about keeping it close to the method. 

> Another, maybe better name for these types would be ‘strategy,’ since that's what it really is.

I think this makes sense. I'll move the type definition and implement what you outlined here.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -17,6 +17,22 @@ import type {CompiledInstruction} from './message';
 export type TransactionSignature = string;
 
+export type TransactionBlockhash = {
+  blockhash?: Blockhash;
+  lastValidBlockHeight?: number;
```
Why optional? Is it not true that, under the new config regime, people _have_ to bring this information in order to be able to use the new implementation of `confirmTransaction`?

marcnjaramillo replied:
I initially tried to keep things similar to other methods on the `Transaction` class where `recentBlockhash` is marked as optional. However, I agree that with the new config options being implemented it doesn't make sense to make these optional. I'll fix this when I refactor it.

steveluscher commented on web3.js/src/message.ts:
```diff
@@ -73,7 +80,16 @@ export class Message {
     this.accountKeys = args.accountKeys.map(account => new PublicKey(account));
-    this.recentBlockhash = args.recentBlockhash;
+    if (
+      // eslint-disable-next-line no-prototype-builtins
```
Oh, sorry if I confused you here. I saw what you wrote about `Object.prototype.hasOwnProperty.call(args, 'latestBlockhash')` and meant that I learned something new, not that you should change it back and suppress the lint warning.

marcnjaramillo replied:
No worries. I couldn't find your comment about this. I can put this back to the way it was before :)

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2775,38 +2778,54 @@ export class Connection {
+
+  // eslint-disable-next-line no-dupe-class-members
+  async confirmTransaction(
+    signature: TransactionSignatureBlockhash | TransactionSignature,
```
See my comment in this review about why this param should probably be named `config` or `strategy` now.

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2815,39 +2834,67 @@ export class Connection {
+          case 'max':
+          case 'root':
+        }
+        setTimeout(() => resolve(TransactionStatus.EXPIRED), timeoutMs);
```
You have the opportunity to `clearTimeout` this once you learn that the transaction is confirmed (ie. `done` gets set to `true`).

marcnjaramillo replied:
Ah yes, I'll add that in as well.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -17,6 +17,22 @@ import type {CompiledInstruction} from './message';
+  lastValidBlockHeight?: number;
+};
+
+export enum TransactionStatus {
+  CONFIRMED = 'CONFIRMED',
+  EXPIRED = 'EXPIRED',
+}
```
If you use `const` enums with no values, TypeScript can optimize them right down to single digit integers.

<img width="904" alt="image" src="https://user-images.githubusercontent.com/13243/165632982-b420b329-0eb7-4387-bd7d-aa7e77fb70ee.png">

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -133,11 +149,38 @@ export type SignaturePubkeyPair = {
+  signatures?: Array<SignaturePubkeyPair>;
   /** A recent blockhash */
-  recentBlockhash?: Blockhash | null;
+  latestBlockhash?: {
```
Why optional. Aren't we saying that if you want to call this new modern method, you _have_ to bring one of these?

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -133,11 +149,38 @@ export type SignaturePubkeyPair = {
+ * List of Transaction object fields that may be initialized at construction
+ */
+export type TransactionNonceCtor = {
   /** Optional nonce information used for offline nonce'd transactions */
   nonceInfo?: NonceInformation | null;
```
Same here. It made sense for these to be optional _before_, because `new Transaction(...)` could be called with any combination of these properties. What you're doing by adding function overloads is adding some rigor around what combination of constructor args are and are not allowed.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -221,11 +269,46 @@ export class Transaction {
+    opts?:
+      | TransactionBlockhashCtor
+      | TransactionNonceCtor
+      | TransactionCtorFields,
```
Still think this type should have a `_DEPRECATED` suffix, to make it super clear.

marcnjaramillo replied:
Okay, I'll add that back in. :) 

Does the rest of it look okay to you?

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -221,11 +269,46 @@ export class Transaction {
+      // eslint-disable-next-line no-prototype-builtins
+    } else if (opts.hasOwnProperty('latestBlockhash')) {
+      const newOpts = opts as TransactionBlockhashCtor;
+      Object.assign(this, newOpts);
```
This has the unfortunate side effect of setting `latestBlockhash` on `this`, whether you like it or not. Consider manually setting every property so that unintended consequences like that don't happen.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -27,23 +47,29 @@ export async function sendAndConfirmTransaction(
 
-  const status = (
+  const signatureBlockhash: TransactionSignatureBlockhash = {
+    signature: signature,
+    blockhash: transaction.recentBlockhash,
+    lastValidBlockHeight: transaction.lastValidBlockHeight,
+  };
+
+  status = (
     await connection.confirmTransaction(
-      signature,
+      signatureBlockhash,
       options && options.commitment,
     )
   ).value;
```
Think of how much easier this will read if you change everything to use the word ‘strategy,’ as suggested above.

marcnjaramillo replied:
Yeah, I agree. It's getting pretty messy to look at right now.

steveluscher reviewed (dismissed):
Alright! We're almost there. Now I'm starting to wonder… does it… work? :)

steveluscher commented:
Tests might look something like this:

```
describe('with a mock clock', () => {
  let clock: SinonFakeTimers;
  beforeEach(() => {
    clock = useFakeTimers();
  });
  afterEach(() => {
    clock.restore();
  });
  it('confirm transaction - timeout', async () => {
    const EXAMPLE_SIGNATURE =
      'LPJ18iiyfz3G1LpNNbcBnBtaS4dVBdPHKrnELqikjER2DcvB4iyTgz43nKQJH3JQAJHuZdM1xVh5Cnc5Hc7LrqC';
    const confirmationPromise =
      connection.confirmTransaction(EXAMPLE_SIGNATURE);
    clock.tick(60 * 1000);
    await expect(confirmationPromise).to.be.rejected;
  });

  it('confirm transaction - block height exceeded', async () => {
    const EXAMPLE_SIGNATURE =
      'LPJ18iiyfz3G1LpNNbcBnBtaS4dVBdPHKrnELqikjER2DcvB4iyTgz43nKQJH3JQAJHuZdM1xVh5Cnc5Hc7LrqC';
    const confirmationPromise = connection.confirmTransaction({
      latestBlockhash: {
        blockhash: 'ABC',
        lastValidBlockheight: 2,
      },
      signature: EXAMPLE_SIGNATURE,
    });
    // Don't give a getSignatureStatus response.
    await mockRpcResponse({
      method: 'getBlockheight',
      params: [],
      value: 0,
    });
    clock.tick(1 * 1000);
    await mockRpcResponse({
      method: 'getBlockheight',
      params: [],
      value: 1,
    });
    clock.tick(1 * 1000);
    await mockRpcResponse({
      method: 'getBlockheight',
      params: [],
      value: 3,
    });
    await expect(confirmationPromise).to.be.rejected;
  });
});
```

steveluscher reviewed (commented):
This isn't a full review; just submitting some comments incrementally!

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
+          case 'max':
+          case 'root':
+        }
+        if (done) clearTimeout();
```
I'll work on fixing this now. Thanks!

> This is one of the reasons that your test doesn't pass right now.

Which test is failing? When I run `yarn test` locally all the tests pass. I'd like to figure out what would be causing the discrepancy.

steveluscher replied:
I fully rebased and found a couple of things I need to fix for _you_ before this test will pass on the very latest.

steveluscher replied:
K. I made it so that you can mock an RPC message that never arrives.

```
await mockRpcMessage({
  method: 'signatureSubscribe',
  params: [blockHeightSignature, {commitment: 'finalized'}],
  result: new Promise(() => {}), // Make the RPC never respond,
});
```

Rebase after #24920 is merged (minutes from now) and you should be able to use that in all three of these new tests.

marcnjaramillo replied:
Ah, okay

steveluscher replied:
Done! Rebase now.

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2825,38 +2835,72 @@ export class Connection {
+    commitment?: Commitment,
+  ): Promise<RpcResponseAndContext<SignatureResult>>;
+
+  /** @deprecated Instead, call `confirmTransaction` using a `TransactionConfirmationConfig` */
```
Missed this one. Fixed it!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
+            }
+            resolve(TransactionStatus.EXPIRED);
+          } catch (error) {
+            reject(error);
```
This is actually something new in your implementation. Previously, the only thing that `confirmTransaction` could really do is to throw an error after the full timeout elapsed. Now, it can fatal _really_ early if the `getBlockHeight` method fatals for any reason:

* connection issues, maybe, or
* the RPC returns a failure response for `getBlockHeight`

While these imply that the block height is not available to the client at the moment, I'm not sure if it implies that the client will _never_ receive the signature confirmation through the websocket. Failure to get the block height might just be transient, and might resolve the very next time you try. Maybe.

So yeah, what do we do? We could:

1. Wait a second and retry. This isn't perfect, because it implies that `expireTx` will last forever if the `checkBlockHeight` method _never_ succeeds.
2. Throw an error if we haven't been able to determine what the block height is after X seconds of trying.
3. Throw an error if we haven't been able to determine what the block height is after X retries.

I'm inclined to suggest that you try #1. It doesn't seem right for a (transient) failure to obtain the blockheight to nerf `confirmTransaction` completely. Let's retry until we get an answer.

cc/ @mvines for thoughts.

marcnjaramillo replied:
Hmm, interesting. I'll get to work on it!

marcnjaramillo replied:
I haven't had a chance to get into this, but I will.

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
+      const outcome = await Promise.race([confirmTx, expireTx]);
+      switch (outcome) {
+        case 'CONFIRMED':
+          console.log('Transaction has been confirmed');
```
If you got here, it's time to `return response`; no `console.log` necessary.

marcnjaramillo replied:
Done!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
+          console.log('Transaction has been confirmed');
+          break;
+        case 'EXPIRED':
+          console.log('Transaction has expired');
```
If you got here, it's time to throw a `TransactionExpiredBlockheightExceededError` or a `TransactionExpiredTimeoutExceededError` depending on the reason for the expiry; no `console.log` necessary.

You may like to resolve with a different outcome depending on the reason for the expiry. That way you can have three branches in your `switch`.

marcnjaramillo replied:
Done!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
       if (subscriptionId) {
         this.removeSignatureListener(subscriptionId);
       }
     }
```
This should be the last line in your function. The `switch` should always return from one of the branches.

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,39 +2909,72 @@ export class Connection {
-      await promiseTimeout(confirmPromise, timeoutMs);
+      const outcome = await Promise.race([confirmTx, expireTx]);
+      switch (outcome) {
+        case 'CONFIRMED':
```
This isn't _really_ `CONFIRMED`. Consider, that `result.err` could be set, meaning that the transaction _failed_. This outcome is more like `PROCESSED` meaning that we successfully _processed_ the transaction and have an answer regarding whether it succeeded or not.

marcnjaramillo replied:
Renamed to `PROCESSED`

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -2,7 +2,26 @@ import {Connection} from '../connection';
+  TransactionSignature,
+} from '../transaction';
+
+class TransactionExpiredBlockheightExceededError extends Error {
+  signature: string;
+
+  constructor(signature: string) {
+    super(`Signature ${signature} has expired.`);
+    this.signature = signature;
+  }
+}
+
+Object.defineProperty(
+  TransactionExpiredBlockheightExceededError.prototype,
+  'name',
+  {
+    value: 'TransactionExpiredBlockheightExceededError',
+  },
+);
```
These should:

* be thrown from `confirmTransaction` when timeouts happen.
* live in `transaction.ts`, or their own file.

marcnjaramillo replied:
Moved to `transaction.ts`

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -903,6 +904,107 @@ describe('Connection', function () {
+      },
+    });
+
+    await expect(blockHeightPromise).to.be.ok;
```
Don't forget to await the promise! This isn't actually testing anything.

Really you want to check something more like this:

```
// Mock stuff *first*
const confirmPromise = connection.confirmTransaction(...);
await expect(confirmPromise).to.eventually.have.property('err', null);
```

Look up the Mocha Chai docs for promise assertions. https://www.chaijs.com/plugins/chai-as-promised/

marcnjaramillo replied:
Fixed tests!

steveluscher reviewed (changes requested):
A lot of cleanup to do here, but I think we can probably merge this in at most 2 more rounds of review if we really go through the list of things left to do carefully!

1. Rebase!
2. Address feedback.
3. Strip out all the `console.logs`.
4. Resolve those `Object.assign` areas in `Transaction`. You still have properties on `Transaction` being set when you don't mean to, and some properties being set twice – once by the `Object.assign` and once on the next line(s).
5. Add that `_DEPRECATED` suffix.
6. Delete code that's commented out.
7. Switch to a const enum.
8. Write tests that assert that the right error instances are thrown from `confirmTransaction` when under both timeout scenarios.

![](https://media2.giphy.com/media/609o8uNjasiJO/giphy.gif?cid=5a38a5a2q7t47sh0z2ehjwi7elkq17x1w5v256q3d7pn2ve9&amp;rid=giphy.gif&amp;ct=g)

steveluscher commented:
Also, time to switch this PR from draft status to ‘ready to review’ status. It's been ready to review for a looong time now :)

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2825,38 +2835,72 @@ export class Connection {
+      | TransactionSignature,
     commitment?: Commitment,
   ): Promise<RpcResponseAndContext<SignatureResult>> {
+    let rawSignature: string = '';
```
I tried just declaring it, but TypeScript complains that it's being used before it's assigned. Any ideas how to get around that?

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -2825,38 +2835,72 @@ export class Connection {
+    } else if (Object.prototype.hasOwnProperty.call(strategy, 'nonce')) {
+      let config = strategy as NonceBasedTransactionConfirmationStrategy;
+      rawSignature = config.nonce;
+    } else if (typeof strategy == 'string') {
```
I tried leaving off the `if` before, but TypeScript complained that `string | BlockHeightBased... | NonceBased.... is not assignable to type 'string'`. I had to move it to the first check to get the error to go away. I also got rid of the `Nonce` things for this PR.

marcnjaramillo commented on web3.js/src/connection.ts:
```diff
@@ -281,6 +280,17 @@ export type RpcResponseAndContext<T> = {
+  lastValidBlockHeight: number;
+};
+
+type NonceBasedTransactionConfirmationStrategy = {
+  signature: TransactionSignature;
+  nonce: string;
+};
```
Okay, done! :)

> the type of this will be `NonceInformation`, right?

Yep!

marcnjaramillo commented:
The test suite passes, but the builds are all failing because of this:

```
> @solana/web3.js@0.0.0-development doc
> set -ex; typedoc --treatWarningsAsErrors

+ typedoc --treatWarningsAsErrors
Warning: BlockheightBasedTransactionConfimationStrategy, defined at src/connection.ts:287, is referenced by Connection.confirmTransaction.confirmTransaction.strategy but not included in the documentation.
ERROR: "doc" exited with 4.
```

I've seen this once before but I have no idea how to fix it.

steveluscher commented:
That build failure just means that you have to write documentation (in the form of code comments) above `BlockheightBasedTransactionConfimationStrategy` so that typedoc can build it into the documentation website for you! Write something that explains when to use that strategy, and a little bit about how it works.

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2825,38 +2834,66 @@ export class Connection {
+    let timeoutId;
     let subscriptionId;
+    let done = false;
     let response: RpcResponseAndContext<SignatureResult> | null = null;
```
There's no need to bind result out here, or to set it inside the subscription callback, because you're _returning_ it now, through the `resolve()`. You can delete it now!

marcnjaramillo replied:
Done!

steveluscher commented on web3.js/src/connection.ts:
```diff
@@ -2865,40 +2902,81 @@ export class Connection {
+      return blockHeight;
+    };
+
+    const expireTx = new Promise<{
+      __type: TransactionStatus.EXPIRED | TransactionStatus.TIMED_OUT;
+      reason: 'blockheight expired' | 'timeout elapsed';
+    }>((resolve, reject) => {
```
Since you went with two different types, you can do this and eliminate the need for a `reason` field!

```suggestion
    const expireTx = new Promise<
      {__type: TransactionStatus.BLOCKHEIGHT_EXCEEDED} | {__type: TransactionStatus.TIMED_OUT}
    >((resolve, reject) => {
```

marcnjaramillo replied:
Done!

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -21,6 +21,12 @@ import type {CompiledInstruction} from './message';
  */
 export type TransactionSignature = string;
 
+export enum TransactionStatus {
```
Using a const enum means the text can get fully erased during compile.

```suggestion
export const enum TransactionStatus {
```

marcnjaramillo replied:
Got it. Updated now.

steveluscher reviewed (commented):
Love it! Just a few quick comments before I grab a bite to eat. I'll finish the review after that!

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -27,23 +45,27 @@ export async function sendAndConfirmTransaction(
 
-  const status = (
+  status = (
     await connection.confirmTransaction(
-      signature,
+      {
+        signature: signature,
+        blockhash: transaction.recentBlockhash!,
+        lastValidBlockHeight: transaction.lastValidBlockHeight!,
+      },
```
If anyone hands you a transaction that uses the old, deprecated call signature, this code will fatal because of a missing `lastValidBlockHeight`.

What you have to do here is to _check_ whether you have what's necessary to make a new-style call or not.

steveluscher replied:
```
const confirmationStrategy = transaction.recentBlockhash != null && transaction.lastValidBlockHeight != null
  // New, blockheight strategy.
  ? {signature, blockhash: transaction.recentBlockhash, lastValidBlockHeight: transaction.lastValidBlockHeight}
  // Old, deprecated style.
  : signature;

status = (
  await connection.confirmTransaction(confirmationStrategy, ...)
).value;
```

marcnjaramillo replied:
Trying to work on this approach. Currently throwing an error because the overloads don't match.

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -77,6 +82,23 @@ const verifySignatureStatus = (
   return status;
 };
 
+class TransactionExpiredBlockheightExceededError extends Error {
+  signature: string;
+
+  constructor(signature: string) {
+    super(`Signature ${signature} has expired.`);
+    this.signature = signature;
+  }
+}
+
+Object.defineProperty(
+  TransactionExpiredBlockheightExceededError.prototype,
+  'name',
+  {
+    value: 'TransactionExpiredBlockheightExceededError',
+  },
+);
+
```
Did it not work to just `import` this error type?

marcnjaramillo replied:
Whoops, I forgot to update this.

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -903,6 +925,132 @@ describe('Connection', function () {
+
+      clock.tick(60 * 1000);
+
+      await expect(timeoutPromise).to.be.rejectedWith(
+        `Signature ${newSignature} has expired: timeout exceeded.`,
+      );
```
The idea is that, since you're throwing an actual custom `Error` with a backing class and everything, you should be able to assert that an error _of that type_ has been thrown from the code – not just that the message matches.

```suggestion
      await expect(timeoutPromise).to.be.rejectedWith(TransactionExpiredTimeoutError)
```

marcnjaramillo replied:
Fixed now!

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -903,6 +925,132 @@ describe('Connection', function () {
+      result: {err: txBlockHeightExpired},
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [],
+      error: txBlockHeightExpired,
+    });
```
We don't want the `getBlockHeight` method to throw; we just want it to advance past `5`.

marcnjaramillo replied:
Updated now.

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -903,6 +925,132 @@ describe('Connection', function () {
+      value: 6,
+    });
+
+    await mockRpcMessage({
+      method: 'signatureSubscribe',
+      params: [blockHeightSignature, {commitment: 'finalized'}],
+      result: {err: txBlockHeightExpired},
+    });
```
This test isn't supposed to test what happens when `signatureSubscribe` returns an error. It's only supposed to test what happens when `signatureSubscribe` _never_ gives a response, and then the `getBlockHeight` message returns `6`.

marcnjaramillo replied:
Fixed this!

steveluscher commented on web3.js/test/connection.test.ts:
```diff
@@ -903,6 +925,132 @@ describe('Connection', function () {
+      value: 4,
+    });
+
+    await mockRpcMessage({
+      method: 'signatureSubscribe',
+      params: [blockHeightSignature, {commitment: 'finalized'}],
+      result: Promise.resolve({err: null}),
+    });
+
+    const blockHeightPromise = connection.confirmTransaction({
+      signature: blockHeightSignature,
+      blockhash: 'sampleBlockhash',
+      lastValidBlockHeight: 6,
+    });
```
Almost! The idea is that these things have to happen in this order in your test:

* you queue up an unresolved promise for the `signatureSubscribe` response
* you queue up some responses for `getBlockHeight`
* you make the `confirmTransaction` call
* you advance time so that every one of the `getBlockHeight` calls gets made
* _then_ you resolve the `signatureSubscribe` promise

That would be a true simulation of receiving a few blockheights before and then receiving a response in the nick of time.

steveluscher reviewed (changes requested):
Nice! Just some tidy up. Let me know if you want help with the tests; I'm happy to check this PR out and give it a go.

marcnjaramillo commented on web3.js/src/transaction.ts:
```diff
@@ -212,11 +248,46 @@ export class Transaction {
+
+  /**
+   * @deprecated `recentBlockhash` has been deprecated.
+   * Please pass `latestBlockhash: {blockHash: BlockHash, lastValidBlockHeight: number}` instead.
```
Hmm, okay. I'll fix that.

marcnjaramillo commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -1,9 +1,27 @@
 import type {Signer} from '../keypair';
 import type {TransactionSignature} from '../transaction';
 
+// export class TransactionExpiredBlockheightExceededError extends Error {
+//   signature: string;
+
+//   constructor(signature: string) {
+//     super(`Signature ${signature} has expired.`);
+//     this.signature = signature;
+//   }
+// }
+
+// Object.defineProperty(
+//   TransactionExpiredBlockheightExceededError.prototype,
+//   'name',
+//   {
+//     value: 'TransactionExpiredBlockheightExceededError',
+//   },
+// );
+
```
Well, darn. I thought I got that one. Done now!

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -27,23 +27,47 @@ export async function sendAndConfirmTransaction(
+        ).value;
+
+  status = (
     await connection.confirmTransaction(
-      signature,
+      {
+        signature: signature,
+        blockhash: transaction.recentBlockhash!,
+        lastValidBlockHeight: transaction.lastValidBlockHeight!,
+      },
       options && options.commitment,
     )
   ).value;
```
Oh oops, I think this one got accidentally left in here?

marcnjaramillo replied:
Yep, I did indeed forget to delete that. Got it now.

steveluscher commented on web3.js/src/util/send-and-confirm-transaction.ts:
```diff
@@ -27,23 +27,47 @@ export async function sendAndConfirmTransaction(
+          await connection.confirmTransaction(
+            {
+              signature: signature,
+              blockhash: transaction.recentBlockhash!,
+              lastValidBlockHeight: transaction.lastValidBlockHeight!,
```
Since you asserted that these are non-null 6 lines above here, you don't need to add the trailing `!` anymore. Remove it, and make sure that TypeScript knows for sure they exist.

marcnjaramillo replied:
Done!

steveluscher reviewed (approved):
Yay! Let's ship this next week!

![](https://media2.giphy.com/media/uLiEXaouJVkuA/200.gif?cid=5a38a5a2sh7ti06unq9ccfr9ot2fz62rj3onxwws3po5djjh&amp;rid=200.gif&amp;ct=g)

Thank you for all of your relentless work on this. Your work here is going to help people get notices of failed transactions _sooner_ so that they can take action _faster_. This is going to ripple all the way through the Solana ecosystem, ultimately increasing the number of _successful_ transactions, and joy using products and services.

I'll do a once over of the tests if you want to fix the few typos, and then I think we're gtg!

steveluscher commented:
And @stellaw1, the same goes for you. Thank you for your contributions here, and for being part of our awesome developer community!

![](https://media4.giphy.com/media/ZmuUzqOIEojjnd4tN8/giphy.gif?cid=5a38a5a2dlgmrbiyyrbuzrl5toliia3ajhlz1gnhrvlx18db&amp;rid=giphy.gif&amp;ct=g)

steveluscher commented:
Here, I fixed up the tests, and made some minor tweaks to let/const bindings. What do you think of these? Feel free to pull them in to this PR or commit them under your account: https://github.com/steveluscher/solana/commits/fix-transaction-confirmation

steveluscher commented:
I made a couple of fixes on https://github.com/steveluscher/solana/commits/fix-transaction-confirmation (I still can't seem to push updates to your branch) but one thing we have to do before shipping this is to make sure that your new tests run with `yarn test:live-with-test-validator`. Right now they don't.

I _think_ the right thing to do here is to simply make them not run when `process.env.TEST_LIVE`; this is probably reasonable because it's more interesting how that code performs under a very specific, simulated scenario, than it is under a realtime scenario.

marcnjaramillo commented:
> I still can't seem to push updates to your branch

I think the only way to fix this is to close this PR and open a new PR from my own personal branch. For whatever reason, I don't have an option to allow maintainer edits on a PR from the `MLH-Fellowship` fork. In the meantime I'll go ahead and merge your changes into this branch.

> I think the right thing to do here is to simply make them not run when process.env.TEST_LIVE; this is probably reasonable because it's more interesting how that code performs under a very specific, simulated scenario, than it is under a realtime scenario.

This makes sense. I'll also make this change.

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -212,11 +248,45 @@ export class Transaction {
+      Object.assign(this, oldOpts);
+      this.recentBlockhash = oldOpts.recentBlockhash;
+      // eslint-disable-next-line no-prototype-builtins
+    } else if (opts.hasOwnProperty('latestBlockhash')) {
```
You meant this to be `lastValidBlockHeight`. It's probably worth writing some tests in `transaction.test.ts` to make sure that the right properties materialize on `Transaction` depending on the shape of the constructor args!

marcnjaramillo replied:
Yep, I know what happened. When I originally wrote this I was using the lastValidBlockhash object rather than just pulling out the values. When I changed that I forgot to change the condition. Thanks for catching it!

steveluscher commented on web3.js/src/transaction.ts:
```diff
@@ -752,9 +800,11 @@ export class Transaction {
   static populate(
     message: Message,
     signatures: Array<string> = [],
+    lastValidBlockHeight?: number,
```
Wait, I'm confused. The `Message` type can have `latestBlockhash` on it. If present, then you have all you need to reconstruct a `Transaction` with a `lastValidBlockHash`. Instead of changing the method sig here, can you just inspect message, looking for `latestBlockhash.lastValidBlockHeight`, and on the `compileMessage` side make sure that you write out a `latestBlockhash`?

marcnjaramillo replied:
That was part of this PR, but we didn't finish it because we realized it was going to be more involved. If you look down further in `Message` to the `from` method, we don't do anything with `lastValidBlockHeight`. So if we include `latestBlockhash` in a constructed message, `lastValidBlockHeight` is `undefined`.
